rgw: add request time into beast log instead of current time

alimaredia opened the pull request:
This would make it easy to calculate the latency
from just that one line

Signed-off-by: Ali Maredia <amaredia@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

cbodley commented:
could you please update the commit message to mention that the common log format (https://httpd.apache.org/docs/current/logs.html#common) specifies that field as "The time that the request was received."?

we don't really conform to its specified format, either. whether you want to tackle that in this PR is up to you

mkogan1 reviewed (approved):
look good

mkogan1 commented:
On my system under load there is a 5 ms diff between the timestamps latency calculation and the latency= log, don't think that it is a deal-breaker.
```
tail -n 25 ./out/radosgw.8000.log | grep "beast:" | awk -F '[ \\[\\]]' '{ cmd_date_done = "date +%s.%N -d " $1 ; cmd_date_start = "date +%s.%N -d " $11 ; DONETIME=0 ; cmd_date_done | getline DONETIME ; close(cmd_date_done) ; STARTTIME=0 ; cmd_date_start | getline STARTTIME ; close(cmd_date_start) ; LATENCY = DONETIME - STARTTIME ; print "done: " $1 " - start: " $11 " = latency: " LATENCY "s" }'
done: 2020-10-13T21:11:35.298+0300 - start: 2020-10-13T21:11:35.289498+0300 = latency: 0.00850201s
done: 2020-10-13T21:11:35.299+0300 - start: 2020-10-13T21:11:35.291498+0300 = latency: 0.00750208s
done: 2020-10-13T21:11:35.299+0300 - start: 2020-10-13T21:11:35.291498+0300 = latency: 0.00750208s
done: 2020-10-13T21:11:35.300+0300 - start: 2020-10-13T21:11:35.291498+0300 = latency: 0.00850201s
done: 2020-10-13T21:11:35.300+0300 - start: 2020-10-13T21:11:35.292498+0300 = latency: 0.00750184s
done: 2020-10-13T21:11:35.301+0300 - start: 2020-10-13T21:11:35.293498+0300 = latency: 0.00750208s
done: 2020-10-13T21:11:35.302+0300 - start: 2020-10-13T21:11:35.293498+0300 = latency: 0.00850201s
done: 2020-10-13T21:11:35.302+0300 - start: 2020-10-13T21:11:35.293498+0300 = latency: 0.00850201s
done: 2020-10-13T21:11:35.303+0300 - start: 2020-10-13T21:11:35.294498+0300 = latency: 0.00850201s
done: 2020-10-13T21:11:35.304+0300 - start: 2020-10-13T21:11:35.295498+0300 = latency: 0.00850201s
done: 2020-10-13T21:11:35.304+0300 - start: 2020-10-13T21:11:35.295498+0300 = latency: 0.00850201s

# compare with regular latency log:
tail -n 25 ./out/radosgw.8000.log | grep "latency="
2020-10-13T21:11:35.298+0300 7f0d2938d700  1 ====== req done req=0x7f0d236f8a60 op status=0 http_status=200 latency=0.009000046s ======
2020-10-13T21:11:35.299+0300 7f0d27b8a700  1 ====== req done req=0x7f0d233f2a60 op status=0 http_status=200 latency=0.008000041s ======
2020-10-13T21:11:35.299+0300 7f0d25385700  1 ====== req done req=0x7f0d23575a60 op status=0 http_status=200 latency=0.008000041s ======
2020-10-13T21:11:35.300+0300 7f0d2a38f700  1 ====== req done req=0x7f0d23b00a60 op status=0 http_status=200 latency=0.009000046s ======
2020-10-13T21:11:35.300+0300 7f0d27b8a700  1 ====== req done req=0x7f0d23473a60 op status=0 http_status=200 latency=0.008000041s ======
2020-10-13T21:11:35.301+0300 7f0d27389700  1 ====== req done req=0x7f0d235f6a60 op status=0 http_status=200 latency=0.008000042s ======
2020-10-13T21:11:35.302+0300 7f0d24b84700  1 ====== req done req=0x7f0d23a7fa60 op status=0 http_status=200 latency=0.009000047s ======
2020-10-13T21:11:35.302+0300 7f0d2a38f700  1 ====== req done req=0x7f0d234f4a60 op status=0 http_status=200 latency=0.009000047s ======
2020-10-13T21:11:35.303+0300 7f0d25b86700  1 ====== req done req=0x7f0d23779a60 op status=0 http_status=200 latency=0.009000047s ======
2020-10-13T21:11:35.304+0300 7f0d26387700  1 ====== req done req=0x7f0d239fea60 op status=0 http_status=200 latency=0.009000046s ======
2020-10-13T21:11:35.304+0300 7f0d2938d700  1 ====== req done req=0x7f0d237faa60 op status=0 http_status=200 latency=0.009000046s ======
```

alimaredia commented:
@mkogan1 @cbodley just added a commit that should change the format of the timestamps from:
`[2020-10-23T12:58:30.725242-0400]`
to
`[10/Oct/2000:13:55:36 -0700]`

mkogan1 commented on src/rgw/rgw_asio_frontend.cc:
```diff
@@ -210,16 +211,22 @@ void handle_connection(boost::asio::io_context& context,
         using ceph::operator<<; // for coarse_real_time
         ldout(cct, 1) << "beast: " << hex << &req << dec << ": "
-            << remote_endpoint.address() << " - - [" << now << "] \""
+            << remote_endpoint.address() << " - - [" << apache_formatted_time << "] \""
```
Looks good, but we need to add sub-second accuracy, possibly like that:
```suggestion
            << remote_endpoint.address() << " - - [" << put_time(local_tm, "%d/%b/%Y:%T.")
            << setfill('0') << setw(3) << std::chrono::duration_cast<std::chrono::milliseconds>(started.time_since_epoch()).count() % 1000
            << put_time(local_tm, " %z") << "] \""
```

mkogan1 replied:
updated awk parser:
```
❯ tail -n 25 ./out/radosgw.8000.log | grep "beast:" | awk -F '[ \\[\\]]' '{ cmd_date_done = "date +%s.%N -d " $1 ; gsub(/\//,"-",$11); sub(/:/," ",$11); cmd_date_start = "date +%s.%N -d \"" $11$12 "\"" ; DONETIME=0 ; cmd_date_done | getline DONETIME ; close(cmd_date_done) ; STARTTIME=0 ; cmd_date_start | getline STARTTIME ; close(cmd_date_start) ; LATENCY = DONETIME - STARTTIME ; print "done: " $1 " - start: " $11$12 " = latency: " LATENCY "s" }'
done: 2020-11-04T12:47:56.334+0200 - start: 04-Nov-2020 12:47:56.322+0200 = latency: 0.0120001s
done: 2020-11-04T12:47:56.335+0200 - start: 04-Nov-2020 12:47:56.324+0200 = latency: 0.0110002s
done: 2020-11-04T12:47:56.337+0200 - start: 04-Nov-2020 12:47:56.322+0200 = latency: 0.0149999s
done: 2020-11-04T12:47:56.338+0200 - start: 04-Nov-2020 12:47:56.325+0200 = latency: 0.013s
done: 2020-11-04T12:47:56.340+0200 - start: 04-Nov-2020 12:47:56.327+0200 = latency: 0.013s
done: 2020-11-04T12:47:56.340+0200 - start: 04-Nov-2020 12:47:56.329+0200 = latency: 0.0109999s
done: 2020-11-04T12:47:56.341+0200 - start: 04-Nov-2020 12:47:56.330+0200 = latency: 0.0110002s
done: 2020-11-04T12:47:56.345+0200 - start: 04-Nov-2020 12:47:56.334+0200 = latency: 0.0109999s
done: 2020-11-04T12:47:56.346+0200 - start: 04-Nov-2020 12:47:56.330+0200 = latency: 0.016s
done: 2020-11-04T12:47:56.348+0200 - start: 04-Nov-2020 12:47:56.336+0200 = latency: 0.0120001s
done: 2020-11-04T12:50:12.938+0200 - start: 04-Nov-2020 12:50:12.935+0200 = latency: 0.00300002s
❯ tail -n 25 ./out/radosgw.8000.log | grep "latency="
2020-11-04T12:47:56.334+0200 7fa56d670700  1 ====== req done req=0x7fa4e735fa70 op status=0 http_status=200 latency=0.011999929s ======
2020-11-04T12:47:56.335+0200 7fa56b66c700  1 ====== req done req=0x7fa4e7563a70 op status=0 http_status=200 latency=0.010999935s ======
2020-11-04T12:47:56.337+0200 7fa57be8d700  1 ====== req done req=0x7fa4e74e2a70 op status=0 http_status=200 latency=0.013999917s ======
2020-11-04T12:47:56.338+0200 7fa54fe35700  1 ====== req done req=0x7fa4e73e0a70 op status=0 http_status=200 latency=0.012999923s ======
2020-11-04T12:47:56.340+0200 7fa575e81700  1 ====== req done req=0x7fa4e725da70 op status=0 http_status=200 latency=0.012999923s ======
2020-11-04T12:47:56.340+0200 7fa565660700  1 ====== req done req=0x7fa4e7461a70 op status=0 http_status=200 latency=0.010999934s ======
2020-11-04T12:47:56.341+0200 7fa556e43700  1 ====== req done req=0x7fa4e71dca70 op status=0 http_status=200 latency=0.010999935s ======
2020-11-04T12:47:56.345+0200 7fa569e69700  1 ====== req done req=0x7fa4e735fa70 op status=0 http_status=200 latency=0.010999935s ======
2020-11-04T12:47:56.346+0200 7fa540e17700  1 ====== req done req=0x7fa4e72dea70 op status=0 http_status=200 latency=0.015999905s ======
2020-11-04T12:47:56.348+0200 7fa55d650700  1 ====== req done req=0x7fa4e7563a70 op status=0 http_status=200 latency=0.011999929s ======
2020-11-04T12:50:12.938+0200 7fa534dff700  1 ====== req done req=0x7fa4e7563a70 op status=0 http_status=200 latency=0.002999982s ======
```

mkogan1 commented:
> On my system under load there is a 5 ms diff between the timestamps latency calculation and the latency= log,

correction - the discrepancy is actually 0.5ms and not 5ms, likely caused by calculation of different accuracy timestamps (msec vs usec) and no such discrepancy in the new suggested implementation example where both timestamps are using msec accuracy.

cbodley commented on src/rgw/rgw_process.cc:
```diff
@@ -305,6 +306,8 @@ int process_request(rgw::sal::RGWRadosStore* const store,
   }
   int op_ret = 0;
+  *user = s->user->get_id().to_str();
+
```
if you're taking the string as a pointer, you need to check for null before assigning to it. `RGWLoadGenProcess::handle_request()` is passing null here (and, if civetweb isn't using the value, it should too)

cbodley commented on src/rgw/rgw_asio_frontend.cc:
```diff
@@ -210,22 +211,30 @@ void handle_connection(boost::asio::io_context& context,
-            << remote_endpoint.address() << " - - [" << now << "] \""
+            << remote_endpoint.address() << " - - [" << put_time(local_tm, "%d/%b/%Y:%T.")
+            << setfill('0') << setw(3) << std::chrono::duration_cast<std::chrono::milliseconds>(started.time_since_epoch()).count() % 1000
+            << put_time(local_tm, " %z") << "] \""
```
i pushed a cleanup commit to https://github.com/cbodley/ceph/commits/wip-pr-37642 that moves this formatting logic into a stream output operator. can you pull that in please?

alimaredia commented:
@cbodley I did the following:
1. added your commit
2. added a `nullptr` check when `user` is assigned in `process_request()`
3. passed `nullptr` into `rgw_civetweb_frontend.cc`'s `process_request()` instead of initializing a string that won't be used. 

I think I'm all set?

cbodley commented on src/rgw/rgw_process.cc:
```diff
@@ -305,6 +306,11 @@ int process_request(rgw::sal::RGWRadosStore* const store,
   int op_ret = 0;
+
+  if(user != nullptr) {
+    *user = s->user->get_id().to_str();
```
`s->user` is also unchecked, and i'm not sure we guarantee that it gets initialized for all requests. maybe something like this?
```c++
if (user && rgw::sal::RGWObject::empty(s->user)) {
  *user = s->user->get_id().to_str();
}
```

dang commented on src/rgw/rgw_process.cc:
```diff
@@ -305,6 +306,11 @@ int process_request(rgw::sal::RGWRadosStore* const store,
   }
   int op_ret = 0;
+
+  if (user && rgw::sal::RGWObject::empty(s->user)) {
```
Should be !rgw::sal::

alimaredia replied:
@dang @cbodley `rgw::sal::RGWObject::empty()` won't take a `s->user` (https://jenkins.ceph.com/job/ceph-pull-requests/63777/consoleFull#779856932a811ea2-3e7b-466b-84b4-d13df7e35809) would `if (user && s->user)` work?

alimaredia replied:
@cbodley if you just run a `curl -L` on the rgw the user that shows up is `anonymous`. Does this imply that `s->user` gets initialized for all requests?

dang replied:
The proper check is:

if (user && !rgw::sal::RGWObject::empty(s->user.get())) {

cbodley commented on src/rgw/rgw_asio_frontend.cc:
```diff
@@ -210,16 +234,17 @@ void handle_connection(boost::asio::io_context& context,
-        ldout(cct, 1) << "beast: " << hex << &req << dec << ": "
-            << remote_endpoint.address() << " - - [" << now << "] \""
+        ldout(cct, 1) << "beast: " << std::hex << &req << std::dec << ": "
+            << remote_endpoint.address() << " - " << user << " [" << log_apache_time{started} << "] \""
```
according to the apache format, this should be a "-" when the user is empty

alimaredia replied:
Will initializing it as a "-" work?

mkogan1 commented:
latency calculation from the timestamps seems correct:
```
2020-11-29T18:34:06.563+0200 7fa359675700  1 beast: 0x7fa24b4581f8: 127.0.0.1 - cosbench [29/Nov/2020:18:34:06.556 +0200] "PUT /b01b000000000000/000000000994 HTTP/1.1" 200 16384 - "aws-sdk-go/1.25.36 (go1.12.13; linux; amd64)" - 
2020-11-29T18:34:06.563+0200 7fa35ce7c700  1 ====== starting new request req=0x7fa24b4581f8 ===== 
2020-11-29T18:34:06.571+0200 7fa34ee60700  1 ====== req done req=0x7fa24b4581f8 op status=0 http_status=200 latency=0.007999953s ====== 
2020-11-29T18:34:06.571+0200 7fa34ee60700  1 beast: 0x7fa24b4581f8: 127.0.0.1 - cosbench [29/Nov/2020:18:34:06.563 +0200] "PUT /b01b000000000000/000000000995 HTTP/1.1" 200 16384 - "aws-sdk-go/1.25.36 (go1.12.13; linux; amd64)" - 
2020-11-29T18:34:06.571+0200 7fa35466b700  1 ====== starting new request req=0x7fa24b4581f8 ===== 
2020-11-29T18:34:06.580+0200 7fa35566d700  1 ====== req done req=0x7fa24b4581f8 op status=0 http_status=200 latency=0.008999947s ====== 
2020-11-29T18:34:06.580+0200 7fa35566d700  1 beast: 0x7fa24b4581f8: 127.0.0.1 - cosbench [29/Nov/2020:18:34:06.571 +0200] "PUT /b01b000000000000/000000000996 HTTP/1.1" 200 16384 - "aws-sdk-go/1.25.36 (go1.12.13; linux; amd64)" - 
2020-11-29T18:34:06.581+0200 7fa34fe62700  1 ====== starting new request req=0x7fa24b4581f8 ===== 
2020-11-29T18:34:06.588+0200 7fa34be5a700  1 ====== req done req=0x7fa24b4581f8 op status=0 http_status=200 latency=0.006999959s ====== 
2020-11-29T18:34:06.588+0200 7fa34be5a700  1 beast: 0x7fa24b4581f8: 127.0.0.1 - cosbench [29/Nov/2020:18:34:06.580 +0200] "PUT /b01b000000000000/000000000997 HTTP/1.1" 200 16384 - "aws-sdk-go/1.25.36 (go1.12.13; linux; amd64)" - 
2020-11-29T18:34:06.588+0200 7fa34b659700  1 ====== starting new request req=0x7fa24b4581f8 ===== 
2020-11-29T18:34:06.596+0200 7fa34a657700  1 ====== req done req=0x7fa24b4581f8 op status=0 http_status=200 latency=0.007999954s ====== 
2020-11-29T18:34:06.596+0200 7fa34a657700  1 beast: 0x7fa24b4581f8: 127.0.0.1 - cosbench [29/Nov/2020:18:34:06.588 +0200] "PUT /b01b000000000000/000000000998 HTTP/1.1" 200 16384 - "aws-sdk-go/1.25.36 (go1.12.13; linux; amd64)" - 
2020-11-29T18:34:06.596+0200 7fa348653700  1 ====== starting new request req=0x7fa24b4581f8 ===== 
2020-11-29T18:34:06.604+0200 7fa344e4c700  1 ====== req done req=0x7fa24b4581f8 op status=0 http_status=200 latency=0.007999953s ====== 
2020-11-29T18:34:06.604+0200 7fa344e4c700  1 beast: 0x7fa24b4581f8: 127.0.0.1 - cosbench [29/Nov/2020:18:34:06.596 +0200] "PUT /b01b000000000000/000000000999 HTTP/1.1" 200 16384 - "aws-sdk-go/1.25.36 (go1.12.13; linux; amd64)" - 

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
tail -n 25 ./out/radosgw.8000.log | grep "beast:" | awk -F '[ \\[\\]]' '{ cmd_date_done = "date +%s.%N -d " $1 ; gsub(/\//,"-",$11); sub(/:/," ",$11); cmd_date_start = "date +%s.%N -d \"" $11$12 "\"" ; DONETIME=0 ; cmd_date_done | getline DONETIME ; close(cmd_date_done) ; STARTTIME=0 ; cmd_date_start | getline STARTTIME ; close(cmd_date_start) ; LATENCY = DONETIME - STARTTIME ; print "done: " $1 " - start: " $11$12 " = latency: " LATENCY "s" }'
done: 2020-11-29T18:34:06.540+0200 - start: 29-Nov-2020 18:34:06.532+0200 = latency: 0.0079999s
done: 2020-11-29T18:34:06.548+0200 - start: 29-Nov-2020 18:34:06.540+0200 = latency: 0.00800014s
done: 2020-11-29T18:34:06.555+0200 - start: 29-Nov-2020 18:34:06.548+0200 = latency: 0.00699997s
done: 2020-11-29T18:34:06.563+0200 - start: 29-Nov-2020 18:34:06.556+0200 = latency: 0.00699997s
done: 2020-11-29T18:34:06.571+0200 - start: 29-Nov-2020 18:34:06.563+0200 = latency: 0.00800014s
done: 2020-11-29T18:34:06.580+0200 - start: 29-Nov-2020 18:34:06.571+0200 = latency: 0.00899982s
done: 2020-11-29T18:34:06.588+0200 - start: 29-Nov-2020 18:34:06.580+0200 = latency: 0.00800014s
done: 2020-11-29T18:34:06.596+0200 - start: 29-Nov-2020 18:34:06.588+0200 = latency: 0.0079999s
done: 2020-11-29T18:34:06.604+0200 - start: 29-Nov-2020 18:34:06.596+0200 = latency: 0.00800014s
```
and anonymous req shows `-` in user designation:
```
2020-11-29T18:47:47.906+0200 7fa341e46700  1 ====== starting new request req=0x7fa24b4581f8 ===== 
2020-11-29T18:47:47.907+0200 7fa33ce3c700  1 ====== req done req=0x7fa24b4581f8 op status=0 http_status=200 latency=0.000999994s ====== 
2020-11-29T18:47:47.907+0200 7fa33ce3c700  1 beast: 0x7fa24b4581f8: 127.0.0.1 - - [29/Nov/2020:18:47:47.906 +0200] "GET /swift/healthcheck HTTP/1.1" 200 0 - "curl/7.69.1" - 

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
curl -sv -D - "http://localhost:8000/swift/healthcheck"
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> GET /swift/healthcheck HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.69.1
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< X-Trans-Id: tx0000000000000000003ef-005fc3d0b3-103f-default
X-Trans-Id: tx0000000000000000003ef-005fc3d0b3-103f-default
< X-Openstack-Request-Id: tx0000000000000000003ef-005fc3d0b3-103f-default
X-Openstack-Request-Id: tx0000000000000000003ef-005fc3d0b3-103f-default
< Content-Type: application/xml
Content-Type: application/xml
< Content-Length: 0
Content-Length: 0
< Date: Sun, 29 Nov 2020 16:47:47 GMT
Date: Sun, 29 Nov 2020 16:47:47 GMT
< Connection: Keep-Alive
Connection: Keep-Alive

< 
* Connection #0 to host localhost left intact
```
lgtm

mkogan1 commented:
pushed fix for Signed-off-by jenkins test error on commit 1be66e5ebb49c425f20cfb458c45989aab686988

mkogan1 commented:
jenkins test make check

mkogan1 commented:
jenkins test make check

mkogan1 commented:
The `make check` test that is failing seems unrelated:
![image](https://user-images.githubusercontent.com/31659604/100764762-5fb05200-33ff-11eb-93a4-72e9c857face.png)

mkogan1 commented:
jenkins retest this please

mkogan1 commented:
jenkins test make check

alimaredia commented:
Link to a successful teuthology run: http://qa-proxy.ceph.com/teuthology/amaredia-2020-12-04_21:25:44-rgw-wip-rgw-apache-log-timestamp-change-distro-basic-smithi/5681019/teuthology.log
